include/rmm/mr/device/cuda_async_memory_resource.hpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include "rmm/cuda_stream_view.hpp"
+
+#include <cuda_runtime_api.h>
+
+#if CUDART_VERSION >= 11020  // 11.2 introduced cudaMallocAsync
+#define CUDA_MALLOC_ASYNC_SUPPORT
+#endif
+
+namespace rmm {
+namespace mr {
+
+/**
+ * @brief `device_memory_resource` derived class that uses `cudaMallocAsync`/`cudaFreeAsync` for
+ * allocation/deallocation.
+ */
+class cuda_async_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Default constructor
+   *
+   * @throws rmm::runtime_error if the CUDA version does not support `cudaMallocAsync`
+   */
+  cuda_async_memory_resource()
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    // Check if cudaMallocAsync Memory pool supported
```

Comment:
I think we want to be creating our own `cudaMemPool_t` with `cudaMemPoolCreate` and setting it as the default pool here. I believe this will be necessary so we can enable IPC.

Reply:
Hmmm, that's debatable... I think the default should be to use the default pool. Then we should provide other constructors with parameters to specify creating a new pool (with options), and/or using a passed pool handle.

Reply:
I like the idea of defaulting to the default pool, though if non-default pools are in use, we can't call `cudaMallocAsync` in `do_allocate` but should instead call `cudaMallocFromPoolAsync`, so I think it's better to just use the latter and always attempt to get & pass the pool handle.

`do_deallocate` is fine, though, because there is no "`cudaFreeFromPoolAsync`", just `cudaFreeAsync`.

Reply:
I was revisiting the doc, and begin wondering if I do `cudaDeviceSetMemPool` first, will the subsequent calls to `cudaMallocAsync` draw memory from the specified pool? The doc isn't clear about this 😢

Reply:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g02bb50d2e2af83e5f24d0b2ab9dadc82
> The memory pool must be local to the specified device. Unless a mempool is specified in the cudaMallocAsync call, cudaMallocAsync allocates from the current mempool of the provided stream's device. By default, a device's current memory pool is its default memory pool.

So looks like it's ok...

Reply:
This is going to require more thought in future PRs because I'm still not sure on this. My current thinking is we should create a new `cudaMemPool_t` held by this resource and not set it as the default but instead just use `cudaMallocFromPoolAsync`.

Reply:
Regardless of whether you need to create a new pool or not (I think it's a design choice up to RMM), my current understanding is that it is indeed the best to keep a `cudaMemPool_t` handle and always call `cudaMallocFromPoolAsync` if it is a concern to you that other applications overwrite the current pool on the device using `cudaDeviceSetMemPool`.